spotipy/json_types.py

```diff
@@ -0,0 +1,573 @@
+    "after": str,
+    "before": str,
+})
+
+Device = TypedDict("Device", {
+    "id": str,
+    "is_active": bool,
+    "is_private_session": bool,
+    "is_restricted": bool,
+    "name": str,
+    "type": str,
+    "volume_percent": int,
+})
+
+Devices = TypedDict("Devices", {
+    "devices": List["Device"],
+})
+
+Disallows = TypedDict("Disallows", {
+    "interrupting_playback": bool,
+    "pausing": bool,
+    "resuming": bool,
+    "seeking": bool,
+    "skipping_next": bool,
+    "skipping_prev": bool,
+    "toggling_repeat_context": bool,
+    "toggling_repeat_track": bool,
+    "toggling_shuffle": bool,
+    "transferring_playback": bool,
+})
+
+Episode = TypedDict("Episode", {
+    "audio_preview_url": str,
+    "description": str,
+    "duration_ms": int,
+    "explicit": bool,
+    "external_urls": "ExternalUrl",
+    "href": str,
+    "html_description": str,
+    "id": str,
+    "images": List["Image"],
+    "is_externally_hosted": bool,
+    "is_playable": bool,
+    "language": str,
+    "languages": List[str],
+    "name": str,
+    "release_date": str,
+    "release_date_precision": str,
+    "restrictions": "EpisodeRestriction",
+    "resume_point": "ResumePoint",
+    "show": "SimplifiedShow",
+    "type": str,
+    "uri": str,
+})
+
+EpisodeRestriction = TypedDict("EpisodeRestriction", {
+    "reason": str,
+})
+
+Error = TypedDict("Error", {
+    "message": str,
+    "status": int,
+})
+
+ExplicitContentSettings = TypedDict("ExplicitContentSettings", {
+    "filter_enabled": bool,
+    "filter_locked": bool,
+})
+
+ExternalId = TypedDict("ExternalId", {
+    "ean": str,
+    "isrc": str,
+    "upc": str,
+})
+
+ExternalUrl = TypedDict("ExternalUrl", {
+    "spotify": str,
+})
+
+Followers = TypedDict("Followers", {
+    "href": str,
+    "total": int,
+})
+
+Image = TypedDict("Image", {
+    "height": int,
+    "url": str,
+    "width": int,
+})
+
+LinkedTrack = TypedDict("LinkedTrack", {
+    "external_urls": "ExternalUrl",
+    "href": str,
+    "id": str,
+    "type": str,
+    "uri": str,
+})
+
+PlayHistory = TypedDict("PlayHistory", {
+    "context": "Context",
+    "played_at": str,
+    "track": "SimplifiedTrack",
+})
+
+PlayerError = TypedDict("PlayerError", {
+    "message": str,
+    "reason": str,
+    "status": int,
+})
+
+Playlist = TypedDict("Playlist", {
+    "collaborative": bool,
+    "description": str,
+    "external_urls": "ExternalUrl",
+    "followers": "Followers",
+    "href": str,
+    "id": str,
+    "images": List["Image"],
+    "name": str,
+    "owner": "PublicUser",
+    "public": bool,
+    "snapshot_id": str,
+    "tracks": List["PlaylistTrack"],
+    "type": str,
+    "uri": str,
+})
+
+PlaylistTrack = TypedDict("PlaylistTrack", {
+    "added_at": str,
+    "added_by": "PublicUser",
+    "is_local": bool,
+    "track": "Track",
+})
+
+PlaylistTracksRef = TypedDict("PlaylistTracksRef", {
+    "href": str,
+    "total": int,
+})
+
+PrivateUser = TypedDict("PrivateUser", {
```

Comment:
The `email`, `explicit_content`, `product`, and `country` properties of the `PrivateUser` object can still be `nil` if certain scopes are missing, and so these should also be optional. See [here](https://developer.spotify.com/documentation/web-api/reference/#object-privateuserobject).

Reply:
There is the same issue that I described [here](https://github.com/plamere/spotipy/issues/439#issuecomment-872464912).